api/clients/accountant.go

```diff
@@ -0,0 +1,154 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	commonpb "github.com/Layr-Labs/eigenda/api/grpc/common"
+	"github.com/Layr-Labs/eigenda/core"
+	"github.com/Layr-Labs/eigenda/core/meterer"
+)
+
+type IAccountant interface {
+	AccountBlob(ctx context.Context, data []byte, quorums []uint8) (uint32, uint64, error)
+}
+
+type Accountant struct {
+	// on-chain states
+	reservation       core.ActiveReservation
+	onDemand          core.OnDemandPayment
+	reservationWindow uint32
+	pricePerSymbol    uint32
+	minNumSymbols     uint32
+
+	// local accounting
+	// contains 3 bins; index 0 for current bin, 1 for next bin, 2 for overflowed bin
+	binUsages         []uint64
+	usageLock         sync.Mutex
+	cumulativePayment *big.Int
+	stopRotation      chan struct{}
+
+	paymentSigner core.PaymentSigner
+}
+
+func NewAccountant(reservation core.ActiveReservation, onDemand core.OnDemandPayment, reservationWindow uint32, pricePerSymbol uint32, minNumSymbols uint32, paymentSigner core.PaymentSigner) *Accountant {
+	//TODO: client storage; currently every instance starts fresh but on-chain or a small store makes more sense
+	// Also client is currently responsible for supplying network params, we need to add RPC in order to be automatic
+	// There's a subsequent PR that handles populating the accountant with on-chain state from the disperser
+	a := Accountant{
+		reservation:       reservation,
+		onDemand:          onDemand,
+		reservationWindow: reservationWindow,
+		pricePerSymbol:    pricePerSymbol,
+		minNumSymbols:     minNumSymbols,
+		binUsages:         []uint64{0, 0, 0},
+		cumulativePayment: big.NewInt(0),
+		stopRotation:      make(chan struct{}),
+		paymentSigner:     paymentSigner,
+	}
+	go a.startBinRotation()
+	// TODO: add a routine to refresh the on-chain state occasionally?
+	return &a
+}
+
+func (a *Accountant) startBinRotation() {
+	ticker := time.NewTicker(time.Duration(a.reservationWindow) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			a.rotateBins()
+		case <-a.stopRotation:
+			return
+		}
+	}
+}
+
+func (a *Accountant) rotateBins() {
+	a.usageLock.Lock()
+	defer a.usageLock.Unlock()
+	// Shift bins: bin_i to bin_{i-1}, set 0 to bin2
+	a.binUsages[0] = a.binUsages[1]
+	a.binUsages[1] = a.binUsages[2]
+	a.binUsages[2] = 0
+}
+
+func (a *Accountant) Stop() {
+	close(a.stopRotation)
+}
+
+// accountant calculates and records payment information
+func (a *Accountant) BlobPaymentInfo(ctx context.Context, dataLength uint64) (uint32, *big.Int, error) {
+	//TODO: do we need to lock the binUsages here in case the blob rotation happens in the middle of the function?
+	// binUsage := a.binUsages[0] + dataLength
+	a.usageLock.Lock()
+	defer a.usageLock.Unlock()
+	a.binUsages[0] += dataLength
```

Comment:
The way that the accountant is doing bin accounting looks to me like it might be incorrect; in particular, it might not line up with the server, since the transitions are happening at different points on the client than on the server. 

I have another idea for how we could possibly do this. Will follow up!

Reply:
updated to use the wrapped bin array